src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -162,14 +162,16 @@ private static string ParseHostNameFromHeader(string hostHeader)
             return hostHeader;
         }
 
-        private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri, bool isProxyConnect)
+        private HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri, bool isProxyConnect)
         {
             Uri uri = request.RequestUri;
+            bool defaultCredentialsUsed = _settings._credentials == _settings._defaultCredentials|| _settings._defaultProxyCredentials == _settings._defaultCredentials || _proxy?.Credentials == _settings._defaultCredentials;
```

Comment:
This code here doesn't really address this comment: https://github.com/dotnet/runtime/pull/303#discussion_r350895987

The idea with @stephentoub comment was that 'defaultCredentialsUsed' could be pre-calculated and stored in the _settings object.  But this code here is still re-calculating it each time.

Reply:
This logic:

```c#
bool defaultCredentialsUsed = _settings._credentials == _settings._defaultCredentials|| _settings._defaultProxyCredentials == _settings._defaultCredentials || _proxy?.Credentials == _settings._defaultCredentials;
```

needs to something more like this:

```c#
bool defaultCredentialsUsedForEitherProxyOrServer;
if (_settings._credentials == CredentialCache.DefaultCredentials)
{
    defaultCredentialsUsedForEitherProxyOrServer = true;
}
else if  (_proxy != null)
{
    if (_proxy.Credentials == CredentialCache.DefaultCredentials)
    {
        defaultCredentialsUsedForEitherProxyOrServer = true;
    }
    else
    {
        defaultCredentialsUsedForEitherProxyOrServer = false;
    }
}
else
{
    if (_defaultProxyCredentials == CredentialCache.DefaultCredentials)
    {
        defaultCredentialsUsedForEitherProxyOrServer = true;
    }
    else
    {
        defaultCredentialsUsedForEitherProxyOrServer = false;
    }
}
```

Feel free to collapse/optimize the syntax above. But this is the logic I think is required.

Reply:
Oh, you are right! I misread the original comment and thought it was only about eliminating property read whereas it's about the whole expression.
Regarding the suggested change in logic. I now agree that my version is incorrect, but I don't quite understand this point in your suggestion.
```C#
else
{
    if (_defaultProxyCredentials == CredentialCache.DefaultCredentials)
    {
        defaultCredentialsUsedForEitherProxyOrServer = true;
    }
    ...
}
```
Why do you set ```defaultCredentialsUsedForEitherProxyOrServer = true``` if ```_proxy``` is null and ```_setting._credentials``` doesn't equal to ```CredentialCache.DefaultCredentials```? I would think in this case we are making a direct call to the server because a **proxy is not set** thus the value of ```defaultCredentialsUsedForEitherProxyOrServer``` must be controlled only by this expression ```_settings._credentials == CredentialCache.DefaultCredentials```.
So, I would write it as follows
```C#
bool defaultCredentialsUsedForEitherProxyOrServer;
if  (_proxy != null)
{
    if (_proxy.Credentials == CredentialCache.DefaultCredentials)
    {
        defaultCredentialsUsedForEitherProxyOrServer = true;
    }
    else if (_proxy.Credentials == null)
    {
        if (_defaultProxyCredentials == CredentialCache.DefaultCredentials)
        {
            defaultCredentialsUsedForEitherProxyOrServer = true;
        }
        else
        {
           defaultCredentialsUsedForEitherProxyOrServer = false;
        }
    }
}
else if (_settings._credentials == CredentialCache.DefaultCredentials)
{
    defaultCredentialsUsedForEitherProxyOrServer = true;
}
```
What do you think?

Reply:
After some more thinking, it seems the logic must be more sophisticated. Having ```_proxy != null``` **doesn't guarantee** that request will be proxied because it might be **bypassed**. Thus, I think we need to rely on the current branching in ```GetConnectionKey``` and check proxy credentials only when the key is being constructed for proxy (i.e. inside if blocks [if (isProxyConnect)](https://github.com/dotnet/runtime/blob/8a01ff8126f185b54b48e72ed53911d367a6b575/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L171) and [if (proxyUri != null)](https://github.com/dotnet/runtime/blob/8a01ff8126f185b54b48e72ed53911d367a6b575/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L192). Otherwise, we must check only ```_settings.credentials``` value.

Reply:
>After some more thinking, it seems the logic must be more sophisticated. Having _proxy != null doesn't guarantee that request will be proxied because it might be bypassed.

Correct.

>Thus, I think we need to rely on the current branching in GetConnectionKey and check proxy credentials only when the key is being constructed for proxy (i.e. inside if blocks if (isProxyConnect) and if (proxyUri != null). Otherwise, we must check only _settings.credentials value.

I think that might be the right approach.

Reply:
It's implemented in the last commit.